sklearn_numba_dpex/common/topk.py

```diff
@@ -0,0 +1,692 @@
+# This implementation takes inspiration from other open-source implementations, such as
+# [1] https://github.com/pytorch/pytorch/blob/master/caffe2/operators/\
+#    top_k_radix_selection.cuh
+# or
+# [2] https://developer.nvidia.com/blog/gpu-pro-tip-fast-histograms-using-shared\
+# -atomics-maxwell
+
+# TODO: apply Dr. TopK optimizations
+
+import math
+from functools import lru_cache
+
+import dpctl.tensor as dpt
+import numba_dpex as dpex
+import numpy as np
+
+from sklearn_numba_dpex.common._utils import (
+    _get_sequential_processing_device,
+    check_power_of_2,
+)
+from sklearn_numba_dpex.common.kernels import (
+    make_initialize_to_zeros_2d_kernel,
+    make_sum_reduction_2d_kernel,
+)
+
+zero_idx = np.int64(0)
+count_one_as_a_long = np.int64(1)
+count_one_as_an_int = np.int32(1)
+
+
+def topk(array_in, k, group_sizes=None):
+    k_in_subset, threshold, work_group_size, dtype, device = _get_topk_threshold(
+        array_in, k, group_sizes
+    )
+    result = dpt.empty(sh=(k,), dtype=dtype, device=device)
+    result_idx = dpt.zeros(sh=(1,), dtype=np.int32)
+    n_threshold_occurences = int(k_in_subset[0])
+    gather_topk_kernel = _make_gather_topk_kernel(
+        array_in.shape[0], k, n_threshold_occurences, work_group_size
+    )
+
+    gather_topk_kernel(array_in, threshold, k_in_subset, result_idx, result)
+    return result
+
+
+def topk_idx(array_in, k, group_sizes=None):
+    k_in_subset, threshold, work_group_size, dtype, device = _get_topk_threshold(
+        array_in, k, group_sizes
+    )
+    result = dpt.empty(sh=(k,), dtype=np.int64, device=device)
+    result_idx = dpt.zeros(sh=(1,), dtype=np.int32)
+    gather_topk_kernel = _make_gather_topk_idx_kernel(
+        array_in.shape[0], k, work_group_size
+    )
+
+    gather_topk_kernel(array_in, threshold, k_in_subset, result_idx, result)
+    return result
```

Comment:
Have I missed something?
```suggestion
    return result_idx
```

Reply:
i'll rename result_idx to index_buffer maybe ?